### DISCUSSION TITLE
Can HW B-Frames be disabled?

### DISCUSSION THREAD

**Discussion by alexd2211** [alvr-org/ALVR#1947] [Q&A]
Tried to use NVENC encoder with GTX 1080 TI but ALVR kept reporting that HW accelerated B Frames are not supported which then results in SteamVR throwing an error cancelling the stream.

Apparently, to disable B frames, one can pass b_ref_mode=0 to ffmpeg to skip B Frames but is this possible to do with ALVR? I've tried a few options (like setting frame strategies, frame counts, etc., to 0) but the error still persisted.

Can one disable B frames or is this a critical features (i.e., if it is not supported, then NVENC cannot be used)? If one can disable B frames, how can I configure ALVR to disable B frames?

Thanks.

**Comment by zmerp** [comment]
We are not enabling B frames. not sure how the error comes up. I guess you are on Linux. Anyone is able to help?

**Reply by alexd2211** [reply]
Hi.

That is correct, I am on Linux (Linux Mint 21.2 to be exact).

Apparently, ffmpeg folks decided to enable HW B frames by default: https://github.com/GyanD/codexffmpeg/issues/78

This means that to use NVENC encoder on hardware that doesn't support HW B-frames, one has to pass the b_ref_mode=0 option to ffmpeg. 

Question then: does ALVR allow one to pass such an option or is this a software change?

Thanks.

**Reply by zmerp** [reply]
@nowrep any comment on this?

**Comment by MrBlackGoo** [comment]
I meet with exactly same problem, and really beg for help.
I have GeForce GTX 1650 and same issue with B Frames appears every time I try to use NVENC codec (ALVR version 20.5.0). I tried to change settings, add some b_ref_mode parameter there, but nothing helps. 
I checked FFMPEG simple encoding of a media file through the console using NVENC and I got exactly same warning ("B frames as references are not supported") and then the error "Provided device doesn't support required NVENC features". But if add "-b_ref_mode 0" encoding works perfectly fine.

After some research I tried to uninstall and recompile ffmpeg from its sources, changing default settings for using B Frames to False. This experiment was successful. It works nice but as long as I actually use "ffmpeg" from the console. ALVR, sadly completely ignore this changes as it keeps throwing same warning and the error every time. I even changed text of the warning, to check if ALVR using it or not, but ALVR didn't show any my changes at all. 
I tried to use ALVR AppImage and tar version (tar version refuse to see headset without "libx264.so.163" in "lib" folder), but both seems don't even use ffmpeg I recompiled. 
My custom ffmpeg located at "/usr/local/bin", and I do not have any other ffmpeg installed (except libs, which I can't remove for multimedia software like "Haruna Media Player" and etc )

What do I miss? 
How to force ALVR do not use B Frames for NVENC codec?

**Reply by zmerp** [reply]
@nowrep @Vixea?

**Reply by MrBlackGoo** [reply]
I am really sorry to annoy you with this, but could anybody from the ALVR's development team please add an option to turn off the usage of B Frames in the NVENC codec settings? A simple switch in the menu would be enough. This would help a lot, and I am sure it would solve the issues for users with old graphic cards. (Or notify if it is already there/will be there)
Thank you!

**Reply by nowrep** [reply]
Does this work?
```diff
diff --git a/alvr/server/cpp/platform/linux/EncodePipelineNvEnc.cpp b/alvr/server/cpp/platform/linux/EncodePipelineNvEnc.cpp
index 47dadd99..96261aa0 100644
--- a/alvr/server/cpp/platform/linux/EncodePipelineNvEnc.cpp
+++ b/alvr/server/cpp/platform/linux/EncodePipelineNvEnc.cpp
@@ -133,6 +133,7 @@ alvr::EncodePipelineNvEnc::EncodePipelineNvEnc(Renderer *render,
     // Delay isn't actually a delay instead its how many surfaces to encode at a time
     av_opt_set_int(encoder_ctx->priv_data, "delay", 1, 0);
     av_opt_set_int(encoder_ctx->priv_data, "forced-idr", 1, 0);
+    av_opt_set_int(encoder_ctx->priv_data, "b_ref_mode", 0, 0);
 
     encoder_ctx->pix_fmt = AV_PIX_FMT_CUDA;
     encoder_ctx->width = width;
```

**Reply by MrBlackGoo** [reply]
What version should I use to try the fix? ALVR v20.6.1?
EDITED: Seems I should wait new version. Will wait patiently.
Thank you for the reply, anyway!
